### ISSUE TITLE
new check: UnnecessarySemicolonInObjectBlock

### ISSUE THREAD

**Issue by strkkk** [checkstyle/checkstyle#6775]
From https://github.com/checkstyle/checkstyle/issues/6752#issuecomment-493911181
Semicolons after code blocks are not necessary and should be detected as violations. It also includes empty statements.
```
$ cat TestClass.java
public class TestClass {

    {/*init block*/}; // violation

    static {}; // violation

    TestClass (){}; // violation

    class A{}; // violation

    void method(boolean cond){
        while(cond){}; // violation
        if(cond){}; // violation
        if(cond){}; // violation
        if(cond); else {}; // violation
        for(;cond;){}; // violation
        {}; // violation
        ; // violation
    }; // violation

    interface aa{}; //violation

    enum aa1{}; // violation

    @interface anno {}; // violation

}; // violation

$ cat conf.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
       <module name="UnnecessarySemicolonInObjectBlock"/>
    </module>
</module>

$ java -jar checkstyle-8.20-all.jar -c conf.xml Test.java
Starting audit...
here is violations at lines commented with // violation
```

Following cases are empty statements and can be ignored with additional property `ignoreEmptyStatements`
```
$ cat EmptyStatements.java
public class EmptyStatements {

    void ignoreMethodEmptyStatements(boolean cond){
      // all unnecessary semis below are ignored, these are empty statements
        while(cond){};
        if(cond){};
        if(cond){};
        if(cond); else {};
        for(;cond;){};
        {};
        ;
    }
}
$ cat conf.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
       <module name="UnnecessarySemicolonInObjectBlock">
           <property name="ignoreEmptyStatements" value="true"/>
        </module>
    </module>
</module>

$ java -jar checkstyle-8.20-all.jar -c conf.xml EmptyStatements.java
Starting audit...
Audit done.
```

**Comment by romani** [comment]
1) UnnecessarySemicolonInObjectBlock ==> UnnecessarySemicolonAfterObjectBlock


2) 
> ignoreEmptyStatements

Do you really think we need property for this. I think we can make such ATTENTION note in documentation to make it clear that to validate. 

I reviewed http://checkstyle.sourceforge.net/config_coding.html#EmptyStatement and its goal is not a unnecessary semicolon, but no operation statements.
Is it a reason why you introduced this property ?

**Comment by rnveach** [comment]
> UnnecessarySemicolonAfterObjectBlock

Why `After` when it is inside the OBJBLOCK via our tree?
````
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|  `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_CLASS -> class [1:7]
|--IDENT -> TestClass [1:13]
`--OBJBLOCK -> OBJBLOCK [1:23]
    |--LCURLY -> { [1:23]
    |--VARIABLE_DEF -> VARIABLE_DEF [2:4]
    |  |--MODIFIERS -> MODIFIERS [2:4]
    |  |--TYPE -> TYPE [2:4]
    |  |  `--LITERAL_INT -> int [2:4]
    |  |--IDENT -> field [2:8]
    |  `--SEMI -> ; [2:13]
    |--SEMI -> ; [2:14]
    `--RCURLY -> } [3:0]
````

> Following cases are empty statements and can be ignored with additional property ignoreEmptyStatements

http://checkstyle.sourceforge.net/config_coding.html#EmptyStatement already does this doesn't it? We shouldn't duplicate the functionality of another check.

**Comment by romani** [comment]
>Why After when it is inside the OBJBLOCK via our tree?

Name of Check should be clear for users who do not aware of our AST nuances.
On such level it is better to not reference AST at all.
Semicolon in code is visually after block.

**Comment by strkkk** [comment]
@romani  @rnveach 
> Semicolon in code is visually after block.

Yes, it looks like semicolon is after block. Moreover, I would say it should be named like `UnnecessarySemicolonAfterBlock` (just Block, not ObjectBlock) because it checks semicolons after method body and init blocks as well.
As for me, 'object block' sounds like a class/interface/enum body. 'Block' is more general.

https://github.com/checkstyle/checkstyle/issues/6752#issuecomment-494862260


> Do you really think we need property for this. I think we can make such ATTENTION note in documentation to make it clear that to validate.

> Is it a reason why you introduced this property ?

I don't think we need it. It is better to keep it simple and only check semicolons after blocks. I added this property after your [comment](https://github.com/checkstyle/checkstyle/issues/6752#issuecomment-494862260), but there must be some misunderstanding.

> I reviewed http://checkstyle.sourceforge.net/config_coding.html#EmptyStatement and its goal is not a unnecessary semicolon, but no operation statements.

The thing is that empty statements and unnecessary semicolons are different from grammar point of view, but for user it is basically same thing - semicolon that can be removed.

**Comment by romani** [comment]
> Moreover, I would say it should be named like UnnecessarySemicolonAfterBlock

yes, much better.

>  but there must be some misunderstanding.

`if (someCondition);` is not the same as `if (someCondition){};` so it is ok for two Checks overlap here a bit.
OK, lets remove this property and let the Check validate all cases.
So when this Check is activated, `if (someCondition){};` will be updated my user to `if (someCondition){}` (no violation), and probably some other Check will violate such code.
Please update description to review overall design/behavior again.

**Comment by rnveach** [comment]
@romani 
> if (someCondition); is not the same as if (someCondition){}; so it is ok for two Checks overlap here a bit.

These are not object blocks and will make the name of the check confusing. If we are going to do all blocks, then check should be named `UnnecessarySemicolonAfterBlock`.

Also if new check is going to do exact same functionality of another check with newer violations, why don't we just add new violations to the old check instead of duplicating it and expanding it?

1 other thing. When we were originally discussing naming this `InObjectBlock` I thought we were going to trigger semicolons after fields too. I haven't seen any mention of fields so far. It seems weird imo that this check is only looking after blocks and not things like:
````
public class TestClass {
    int field;; // violation - double ;s
}
enum TestEnum {
    NAME;; // violation - double ;s
}
````
This is why I thought we were going with `InObjectBlock` because we are searching inside the {}s of the TestClass' block, which we call the object block.

**Comment by strkkk** [comment]
> I thought we were going to trigger semicolons after fields too. I haven't seen any mention of fields so far. It seems weird imo that this check is only looking after blocks and not things like

I was sure that these were empty statements and should be detected by `EmptyStatementCheck`

May be, we should consider additional logic for EmptyStatementCheck or it is some parser bug (if it is not an empty statement, what it is?)

> `if (someCondition){};` will be updated my user to `if (someCondition){}` (no violation), and probably some other Check will violate such code.

There is already `EmptyStatementCheck` that covers cases with if/else/etc. Basically, it covers unnecessary semicolons inside method/constructor/init blocks body.

This check can cover either only semicolons after blocks, or all unnecessary semicolons that are outside method/constructor/init blocks body.

**Comment by romani** [comment]
@strkkk , I agree with you.
EmptyStatement is separate target, some other Check should do this.

Discussion become too big and it is hard to maintain it. I recommend close this issue, make summary(all corrections applied) from it and make new issue. @strkkk , please do.
We will invite team to approve new Check design in new issue.

**Comment by strkkk** [comment]
@romani 

I do not really have a clear picture what should be summarized.
Please answer:
1. What is a name of the check
2. Do I understand correctly that it should cover cases that are not covered by `EmptyStatementCheck`? and no additional property needed.

**Comment by romani** [comment]
> What is a name of the check

UnnecessarySemicolonAfterBlock

> Do I understand correctly that it should cover cases that are not covered by EmptyStatementCheck? 

I am not sure I understand. UnnecessarySemicolonAfterBlock care only about `};` without any consideration of content of block.

-----

Please create new issue with definition of behavior again, to make last review before approval.

**Comment by strkkk** [comment]
> I am not sure I understand. UnnecessarySemicolonAfterBlock care only about `};` without any consideration of content of block.

@romani 
Created #6847
Cases inside method is already covered by `EmptyStatementCheck`, the question is - should it duplicate its functionality or not.
Another thing is fields mentioned in https://github.com/checkstyle/checkstyle/issues/6775#issuecomment-496278830
Should empty statement check be extended to cover such cases?

**Comment by romani** [comment]
>the question is - should it duplicate its functionality or not.

It should not

>int field;;

This should be reported by EmptyStatementCheck. Please open issue on this.

**Comment by strkkk** [comment]
created #6849
